### PR TITLE
Use a magic naming scheme to locate the initial root metadata

### PR DESCRIPTION
**Pull request by phenaproxima**
This PR implements the idea @davidstrauss came up with in #18: to use a magic naming scheme to locate the initial root metadata.

The idea is that we expect the initial root metadata to be named after the repository URL, with the various components separated by a period. So `http://localhost:8080` becomes `http.localhost.8080.json`. Subdirectories are considered too, so `https://packages.drupal.org/8` would become `https.packages.drupal.org.8.json`. The file is expected to be in a `tuf` directory next to the active `composer.json`.

Honestly, I like this approach better than #18. We could use the Drupal scaffold plugin to automatically ship this file into the correct location, and the magic naming is (IMHO) pretty easy to grok. This doesn't need to call out to the Internet, so it's safer, and it actually _removes_ configuration from `composer.json`, keeping it simple and self-sufficient.

### CODE REVIEW

File: src/TufValidatedComposerRepository.php [hunk 1]
```diff
@@ -62,10 +58,12 @@ public function __construct(array $repoConfig, IOInterface $io, Config $config,
...
-                }
+                $sourcePath = implode(DIRECTORY_SEPARATOR, [
+                    dirname($config->getConfigSource()->getName()),
+                    'tuf',
+                    $repoKey,
+                ]);
+                $fs->copy("$sourcePath.json", $rootFile);
             }
 
             // For unit testing purposes, allow the updater instance to be provided
```
Comment (tedbow) [inline src/TufValidatedComposerRepository.php:69]: No change but just mention I am not sure this comment still makes sense

**Review by davidstrauss — COMMENTED** [review]
I think including the scheme (`http`) is possibly overkill, especially when we can always introduce discovery that allows a more specific name to be preferred over a less specific one. For example, we could always update the implementation to look for `https.packages.drupal.org.8.json` before looking for `packages.drupal.org.8.json`.

One risk -- that might be negligibly narrow given how the files in this directory will be controlled, is collision. `https://packages.drupal.org/8080` and `https://packages.drupal.org:8080/` would seem to collide. Is that a problem? Solutions might be (1) always include the port, (2) never include the port, (3) conditionally include the port but in a way that can't collide with the top-level part of the path.

I haven't reviewed the code, just the proposed naming.

File: src/TufValidatedComposerRepository.php (hunk 1 above)
Comment (phenaproxima) [reply src/TufValidatedComposerRepository.php]: Fixed.

**Comment by phenaproxima** [comment]
@davidstrauss, I included the scheme because I figured it helps to disambiguate and unique-ify the identifier. It's highly unlikely that the `https` and `http` versions of a particular host are somehow two different repositories, but not impossible. I'd be totally comfortable removing it if you feel it's too much.

I like idea of including the port, even if it's not specified. Either `parse_url($url, PHP_URL_PORT)` returns a port number, or we default it to 80 (for HTTP) or 443 (for HTTPS). Thoughts?

**Comment by phenaproxima** [comment]
One thought: in a follow-up, we could introduce optional configuration for this plugin which allows an advanced case to map a repo URL to a specific root file, thus bypassing magic naming. I picture something like this:

```
"extra": {
  "tuf": {
    "root-files": {
      "https://packages.drupal.org": "drupal.org.secure.json",
      "http://packages.drupal.org": "drupal.org.insecure.json"
    }
  }
}
```

File: README.md [hunk 2]
```diff
@@ -18,6 +18,13 @@ The plugin examines `composer` type repositories. For any that contain an additi
 during package discovery and download operations, validating that the repository and package are not being tampered
 with.
 
+In accordance with the [TUF specification](https://github.com/theupdateframework/specification/blob/v1.0.9/tuf-spec.md#5-detailed-workflows),
+projects using this plugin must supply a set of trusted keys for each repository they want to protect with TUF. Each
+TUF-protected repository should provide a JSON file named for the repository URL, with non-alphanumeric characters
+replaced by a single period. For example, if the repository URL is `https://repo.example.net:1234/packages`, then the
+key file should be called `https.repo.example.net.1234.packages.json`. All key files must be stored in a directory
```
Comment (tedbow) [inline README.md:25]: I don't think we can just replace everything with `.`

`site.coop.info/packages`  and `site.coop/info/packages` would both be converted to `site.coop.info.packages` 
so probably need to either have folder everything before first `/` or replace `/` with something else. maybe `-`

**Comment by davidstrauss** [comment]
> So maybe never including the port is the right idea -- but that could certainly exclude certain exotic set-ups.

I figured the port-based naming is mostly useful for development. I don't particularly like how it also implies the scheme (80 vs. 443) in production cases. We could look for a file matching the host+port first and then fall back to one that only has the host. This would support development and exotic use cases, never collide (if we do it right), and keep the standard case clean.

> in a follow-up, we could introduce optional configuration for this plugin which allows an advanced case to map a repo URL to a specific root file, thus bypassing magic naming

My mantra here is "convention over configuration over code," and I'd prefer we avoid configurability (or, worse, the ability to custom-code how it does discovery, thought that isn't proposed) unless we encounter a use case that expressly requires it. I also like to hunt for ways to accommodate edge cases by extending the system of convention, at least where possible.

For example, we could offer a hash of the entire repo URL as the first-attempted way to discover the right `root.json`, thus allowing, by convention, the ability to pin a `root.json` file to a specific URL string.

In conclusion, I'd probably do discovery this way (descending priority of preference):

1. `sha256($repourl).json`
2. `$fqdn.json`

The chance for collision is so negligible that I think we can ignore it.

**Comment by phenaproxima** [comment]
That sounds reasonable to me. I'd suggest that, to support cases like `packages.drupal.org/7` an `packages.drupal.org/8` (which, as we know, already exist), we do this:

1. `sha256($url).json`
2. `$fqdn-$path`.json
3. `$fqdn.json`

What do you think?

File: README.md (hunk 2 above)
Comment (phenaproxima) [reply README.md:25]: Yeah, you're right...I suspect we are going to use a somewhat different scheme (discussing with @davidstrauss below, so marking this as resolved).

**Comment by davidstrauss** [comment]
> That sounds reasonable to me. I'd suggest that, to support cases like packages.drupal.org/7 an packages.drupal.org/8 (which, as we know, already exist)

It's news to me if we need separate root data for 7 and 8. Is this a speculated use case? We should be able to accomplish any separations of control by delegation if we need to isolate signing authority, which would still share the same root.

**Comment by phenaproxima** [comment]
> It's news to me if we need separate root data for 7 and 8.

Oh, I don't think we'll need separate root data for those two _particular_ repositories. I just wonder if, for cases similar to that, it might make sense to allow the path in there.

**Comment by davidstrauss** [comment]
> ANOTHER EDIT: Although, using a hash of the URL might actually be more brittle, since http://localhost:8080/ and http://localhost:8080 will hash differently

I didn't intend to get into details, but it's possible to canonicalize the URL using a pattern to avoid some brittleness (at the expense of the convention being more complex than a hash of the URL string). I think there are merits to each approach. We also don't really avoid the canonicalization question simply by re-packing the data into a file name. For example, does `https://example.com/one/` have a path with an empty string as the final part, or do we just treat "/one/" and "/one" equivalently (which many servers do, but isn't really RFC).

I do think pure FQDN has elegance and aligns our security namespace granularity with the expectations people have from TLS. I agree that we can always introduce a finer-grained option later if we think FQDN meets immediate needs.

I'm probably as guilty of bikeshedding here, too, but I do like to defend simplicity. :-)

**Comment by phenaproxima** [comment]
Fair enough, @davidstrauss! I think you've won me over. I changed it to search for a SHA of the URL, followed by the FQDN, in that order. (I also did some other refactoring to make the code a little simpler...hopefully.)

File: src/ComposerFileStorage.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+     *   The current Composer configuration.
+     *
+     * @return static
+     *   The storage object.
+     */
+    public static function create(string $url, Config $config): self
+    {
+        $basePath = implode(DIRECTORY_SEPARATOR, [
+            static::basePath($config),
+            preg_replace('/[^[:alnum:]\.]/', '-', $url),
```
Comment (tedbow) [inline src/ComposerFileStorage.php:56]: This will still have same problem

> I don't think we can just replace everything with .
> 
> site.coop.info/packages and site.coop/info/packages would both be converted to site.coop.info.packages
> so probably need to either have folder everything before first / or replace / with something else. maybe -

now it would just both be `site-coop-info-packages`. The point I was trying to make was avoid this collisions you need to replace `.` and `/` with different characters. Not the same character.

Also because the example domains that are conflicting are `coop.info` and `site.coop` this a collision that could between organization not something 1 organization could solve by using a different subdomain or path.  

I person could have repo from `site.coop.info/packages` and then later the `site.coop/info/packages` and they could pick up the wrong root.json unless I missing something

**Review by tedbow — CHANGES_REQUESTED** [review]
Still have the possible confict with different domains resolving to the same initial root.json

File: src/ComposerFileStorage.php (hunk 3 above)
Comment (phenaproxima) [reply src/ComposerFileStorage.php:56]: Good point -- but this is actually what we had before this PR too. See https://github.com/php-tuf/composer-integration/blob/main/src/TufValidatedComposerRepository.php#L54. (In fact, this dates all the way back to before I even started working on this plugin.) So IMHO we should open a follow-up issue to work on that and refine it, but not block on this. What do you think?

Comment (phenaproxima) [reply src/ComposerFileStorage.php:56]: Or, since we're talking about a persistent storage directory, which is used under the hood, we could absolutely just switch to using a SHA-256 hash, which would get around this.

Comment (phenaproxima) [reply src/ComposerFileStorage.php:56]: Another point to consider: this regex (or something very similar) is, I believe, used by Composer when building cache directories for package metadata: https://github.com/composer/composer/blob/master/src/Composer/Repository/ComposerRepository.php#L134

So if they're not worried about collisions, I'm not either. That said, I'd rather just switch to using a hash of the URL as the directory name, so we can put this collision issue to bed forever (or at least until SHA-256 is broken :))

**Comment by phenaproxima** [comment]
> Still have the possible confict with different domains resolving to the same initial root.json

How? Right now, the initial root.json can be named in two ways: the SHA hash of the full URL string, or the domain name. I don't see how that will cause a collision.

File: src/ComposerFileStorage.php (hunk 3 above)
Comment (tedbow) [reply src/ComposerFileStorage.php]: to me using the hash would obscure it more than necessary. say if a new root.json was add to an existing if the directory was something meaningful then it would easier to understand look at changes in git or just looking at the directory structure. With the hash it would not be clear which repository the root.json.

Yeah lets make a follow up to solve this. It should a 0.1.0 or at least 1.0.0 blocker as this not something want to change after people are using this.

Comment (phenaproxima) [reply src/ComposerFileStorage.php:56]: Opened #24 to discuss/solve this.

**Review by tedbow — APPROVED** [review]
Looks good. thanks for the discussion and the changes!

**Comment by davidstrauss** [comment]
Just in case anyone stumbles on this discussion later, I renamed this PR to call it a "naming convention" so that we can avoid the word "magic," which is a term that sets off alarms in my head that aren't actually applicable to what we're doing here.
